src/main/java/net/minecraftforge/event/ModMismatchEvent.java

```diff
@@ -0,0 +1,184 @@
+ * Fires when the mod loader is in the process of loading a world that was last saved
+ * with mod versions that differ from the currently-loaded versions. This can be used to
+ * enqueue work to run at a later point, such as multi-file migration of data.
+ * </p>
+ *
+ * <p>
+ * <b>Note that level and world information has not yet been fully loaded;</b> as such, it is
+ * unsafe to access server or level information during handling of this event.
+ * </p>
+ *
+ * <p>This event is not {@linkplain Cancelable cancellable}, and does not {@linkplain HasResult have a result}.</p>
+ * <p>This event is fired on the {@linkplain FMLJavaModLoadingContext#getModEventBus() mod-specific event bus},
+ * on both {@linkplain LogicalSide logical sides}.</p>
+ */
+public class ModMismatchEvent extends Event implements IModBusEvent
+{
+    /**
+     * The level being loaded. Useful for things like {@link net.minecraft.world.level.storage.DimensionDataStorage}
+     * to manage multiple files changing between mod versions.
+     */
+    @Nullable
+    private final LevelStorageSource.LevelDirectory levelDirectory;
+
+    /**
+     * A set of previously-known versions that have mismatched with the currently loaded versions.
+     */
+    private final HashMap<String, MismatchedVersionInfo> versionDifferences;
+
+    /**
+     * Which mods have specified that they have handled version mismatches.
+     */
+    private final HashMap<String, ModContainer> resolved;
+
+    @ApiStatus.Internal
+    public ModMismatchEvent(@Nullable LevelStorageSource.LevelDirectory levelDirectory, Map<String, ArtifactVersion> previousVersions, Map<String, ArtifactVersion> missingVersions)
+    {
+        this.levelDirectory = levelDirectory;
+        this.resolved = new HashMap<>(previousVersions.size());
+        this.versionDifferences = new HashMap<>();
+        previousVersions.forEach((modId, version) -> versionDifferences.put(modId, new MismatchedVersionInfo(version, ModList.get()
+                .getModContainerById(modId)
+                .map(ModContainer::getModInfo)
+                .map(IModInfo::getVersion)
+                .orElse(null))));
+
+        missingVersions.forEach((modId, version) -> versionDifferences.put(modId, new MismatchedVersionInfo(version, null)));
+    }
+
+    /**
+     * Gets the current level directory for the world being loaded.
+     * Can be used for file operations and manual modification of mod files before world load.
+     */
+    @Nullable
+    public LevelStorageSource.LevelDirectory getLevelDirectory()
+    {
+        return this.levelDirectory;
+    }
+
+    /**
+     * Fetch a previous version of a given mod, if it has been mismatched.
+     * @param modId The mod to fetch previous version for.
+     * @return The previously known mod version, or {@link Optional#empty()} if unknown/not found.
+     */
+    @Nullable
+    public ArtifactVersion getPreviousVersion(String modId)
+    {
+        if(this.versionDifferences.containsKey(modId))
```

Comment:
Formatting: space between `if` keyword and parenthesis.
```suggestion
        if (this.versionDifferences.containsKey(modId))
```